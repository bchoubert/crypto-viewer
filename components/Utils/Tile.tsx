import React, { FC, ReactNode, useMemo } from 'react';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';

import Colors from '../../assets/Colors';

export enum TileMode {
  CLEAR = 'CLEAR',
  LIGHT = 'LIGHT',
  FULL = 'FULL',
  TALL = 'TALL',
};

const styles = StyleSheet.create({
  tile: {
    flex: 1,
    margin: 10,
    height: 90,
    backgroundColor: Colors.lightGray,
    borderRadius: 10,
    display: 'flex',
    flexDirection: 'column',
    padding: 10
  },
  tall: {
    height: 180,
  },
  tile_number: {
    flex: 1,
    fontSize: 28
  },
  tile_long_number: {
    flex: 1,
    fontSize: 25
  },
  tile_label: {
    fontSize: 14
  }
});

const stylesForTileMode = {
  [TileMode.CLEAR]: styles.tile,
  [TileMode.LIGHT]: styles.tile,
  [TileMode.FULL]: styles.tile,
  [TileMode.TALL]: { ...styles.tile, ...styles.tall },
}

interface TileProps {
  mode: TileMode,
  label: string | ReactNode;
  number: string | ReactNode;
  color: string;
  isLongNumber?: boolean;
  style?: object,
  onPress?: () => void;
}

const Tile: FC<TileProps> = ({
  mode,
  label,
  number,
  color,
  isLongNumber,
  style,
  onPress,
}) => {

  const backgroundColor = useMemo(
    () => {
      switch(mode) {
        case TileMode.CLEAR:
          return Colors.white;
        case TileMode.LIGHT:
          return `${color}22`;
        case TileMode.FULL:
        case TileMode.TALL:
          return color;
      }
    },
    [color, mode],
  );

  const textColor = useMemo(
    () => {
      switch(mode) {
        case TileMode.CLEAR:
          return Colors.darkGray;
        case TileMode.LIGHT:
          return Colors.darkGray;
        case TileMode.FULL:
        case TileMode.TALL:
          return Colors.white;
      }
    },
    [color, mode],
  );

  return (
    <TouchableOpacity style={{ ...stylesForTileMode[mode], backgroundColor, ...(style || {}) }} onPress={onPress}>
      <Text style={{
        ...(isLongNumber ? styles.tile_long_number : styles.tile_number),
        color: textColor,
      }}>
        {number}
      </Text>
      {(typeof label === 'string') ? (
        <Text style={{ ...styles.tile_label, color: textColor }}>
          {label}
        </Text>
      ) : (
        <View>{label}</View>
      )}
    </TouchableOpacity>
  );
}

export default Tile;
