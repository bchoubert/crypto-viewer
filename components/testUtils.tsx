/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
