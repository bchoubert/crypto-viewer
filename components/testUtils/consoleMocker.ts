/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
