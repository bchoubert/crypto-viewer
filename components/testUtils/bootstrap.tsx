/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
