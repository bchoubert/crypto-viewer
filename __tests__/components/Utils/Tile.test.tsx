import React from 'react';
import { Text } from 'react-native';
import { renderNode } from '../../../components/testUtils/bootstrap';
import Tile, { TileMode } from '../../../components/Utils/Tile';

describe('<Tile />', () => {
  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <Tile
        mode={TileMode.FULL}
        label="Label"
        number="5.52"
        color="#335588"
      />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Tile modes', () => {
    const commonProps = {
      label: 'Label',
      number: '5.52',
      color: '#335588',
    };

    const clearNode = renderNode(
      <Tile
        {...commonProps}
        mode={TileMode.CLEAR}
      />,
    );

    clearNode.getByText(commonProps.number);
    clearNode.getByText(commonProps.label);

    const lightNode = renderNode(
      <Tile
        {...commonProps}
        mode={TileMode.LIGHT}
      />,
    );

    lightNode.getByText(commonProps.number);
    lightNode.getByText(commonProps.label);

    const fullNode = renderNode(
      <Tile
        {...commonProps}
        mode={TileMode.FULL}
      />,
    );

    fullNode.getByText(commonProps.number);
    fullNode.getByText(commonProps.label);

    const tallNode = renderNode(
      <Tile
        {...commonProps}
        mode={TileMode.TALL}
      />,
    );

    tallNode.getByText(commonProps.number);
    tallNode.getByText(commonProps.label);
  });

  it('Label node', () => {
    const labelToTest = 'LABEL_TO_TEST';

    const node = renderNode(
      <Tile
        mode={TileMode.LIGHT}
        number="5.52"
        label={(
          <Text>{labelToTest}</Text>
        )}
        color="#335588"
      />,
    );

    node.getByText(labelToTest);
  });

  it('Long number mode', () => {
    const numberToTest = '5.5264548616768761';

    const node = renderNode(
      <Tile
        label="Label"
        color="#335588"
        mode={TileMode.LIGHT}
        number={numberToTest}
        isLongNumber
      />,
    );

    node.getByText(numberToTest);
  });
});
