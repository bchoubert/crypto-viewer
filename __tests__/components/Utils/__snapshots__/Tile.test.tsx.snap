// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> Prints correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#335588",
      "borderRadius": 10,
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "height": 90,
      "margin": 10,
      "opacity": 1,
      "padding": 10,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "flex": 1,
        "fontSize": 28,
      }
    }
  >
    5.52
  </Text>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 14,
      }
    }
  >
    Label
  </Text>
</View>
`;
